### DISCUSSION TITLE
Frozen Collections

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#90059] [Q&A]
1.  Can STJ deserialize into these?
2.  Is there a FrozenList?

**Comment by huoyaoyuan** [comment]
> Can STJ deserialize into these?

We can add converters for them in the future, of course. You can also write yours.

> Is there a FrozenList?

It's just an `ImmutableArray`. The frozen collections are optimized for reading.

**Reply by huoyaoyuan** [reply]
The namings are actually a bit confusing. Immutable- lists and dictionaries are optimized for incremental creating. We really need a documentation for "choosing right collection". @dotnet/area-system-collections

**Reply by TonyValenti** [reply]
The problem with immutablearray is that it is a struct and not a class. 

Can you put in a request to have deserializers made for the frozen classes?

We currently use immutables and records for all our json interop but I think frozens would be more performant.

**Reply by stephentoub** [reply]
> We currently use immutables and records for all our json interop but I think frozens would be more performant.

Frozen collections are slow to construct. They're intended for situations where you create them rarely and then read from them a lot. We purposefully haven't added built in converters for them because using them in an object model that's frequently deserialized is very likely the wrong choice.

**Reply by stephentoub** [reply]
> The problem with immutablearray is that it is a struct and not a class.

Why is that a problem? If you need a reference, you can box it. It implements both IList and IReadOnlyList. You can also wrap a ReadOnlyCollection around an array.

**Reply by TonyValenti** [reply]
@stephentoub -
I'll take a look at ways we can use immutable array. 

Regarding performance for frozen collections -
In our use case, we deserialize the json once, cache it in memory, and then query across it many times. Based on what you described, the frozen collections actually would be the ideal classes for us.

In our scenario, we want immutable + ultra fast read performance and we are ok sacrificing creation performance.

**Reply by stephentoub** [reply]
Can't you just deserialize it as an array/dictionary then and create the frozen dictionary/set from it? That's exactly what a converter would, and it's one line of code.

**Reply by TonyValenti** [reply]
I can't because the dictionaries are not the root object - they are  either properties on the root or sub properties on child objects.

**Reply by stephentoub** [reply]
I don't understand. What type are you deserializing today? Just create the frozen collection from that.

**Reply by TonyValenti** [reply]
Hi @stephentoub -
When we get Json from APIs, we generally get something that looks like this:
```json
{
  "Data": [
    {
      "Name": "John Smith",
      "Custom_Values": {
        "14": {
          "Definition": {
            "Id": 46
          },
          "Value": "2023-17-16",
          "Id": 7,
          "Created_At": "2023-08-05T18:15:20.5843392+00:00",
          "Updated_At": "2023-08-05T18:15:20.5843394+00:00"
        },
        "26": {
          "Definition": {
            "Id": 37
          },
          "Value": "Motion to Quash",
          "Id": 9,
          "Created_At": "2023-08-05T18:15:20.5843659+00:00",
          "Updated_At": "2023-08-05T18:15:20.5843659+00:00"
        }
      },
      "Id": 1,
      "Created_At": "2023-08-05T18:15:20.5838881+00:00",
      "Updated_At": "2023-08-05T18:15:20.5839574+00:00"
    },
    {
      "Name": "Jane Doe",
      "Custom_Values": {
        "14": {
          "Definition": {
            "Id": 46
          },
          "Value": "2001-02-02",
          "Id": 79,
          "Created_At": "2023-08-05T18:15:20.5947717+00:00",
          "Updated_At": "2023-08-05T18:15:20.5947718+00:00"
        },
        "27": {
          "Definition": {
            "Id": 456
          },
          "Value": "Wife",
          "Id": 123,
          "Created_At": "2023-08-05T18:15:20.5947739+00:00",
          "Updated_At": "2023-08-05T18:15:20.5947739+00:00"
        }
      },
      "Id": 1,
      "Created_At": "2023-08-05T18:15:20.5947695+00:00",
      "Updated_At": "2023-08-05T18:15:20.5947698+00:00"
    }
  ],
  "Meta": {}
}
```

We will have classes like this:
```csharp

    public record ResponseJson { 
        public MetaJson? Meta { get; init; }
    }

    public record ResponseJson<T> : ResponseJson { 
        public T? Data { get; init; }
    }

    public record MetaJson { 
        //...
    }

    public record EntityJson { 
        public long Id { get; init; }
        public DateTimeOffset Created_At { get; init; }
        public DateTimeOffset Updated_At { get; init; }
    }

    public record PersonJson : EntityJson { 
        public string? Name { get; init; }
        public ImmutableDictionary<long, CustomValueJson>? Custom_Values { get; init; } // <--- This is a high access property
    }

    public record CustomValueJson : EntityJson { 
        public IdReferenceJson<CustomDefinitionJson>? Definition { get; init; }
        public string? Value { get; init; }
    }

    public record IdReferenceJson { 
        public long Id { get; init; }
    }

    public record IdReferenceJson<T> : IdReferenceJson { 
        
    }

    public record CustomDefinitionJson : EntityJson { 
        //...
    }

```

And then we'll deserialize with something like this:
```
var R = System.Text.Json.JsonSerializer.Deserialize<ResponseJson<PersonJson>>(Content);
```

And we'll cache the individual elements in ```R.Data``` for the entire duration of the application.  During execution we will make extensive access to the values in the ```Custom_Values``` member.  It is an ```ImmutableDictionary``` which is great because it enforces that the values cannot change.  However, given the performance benefits of frozen collections (and the fact that we never change or mutate it), I would like to leverage FrozenDictionary instead.

The above example is concise, but during runtime it is not uncommon for use to have 200k ```PersonJson``` instances each which has 50-200 entries in ```Custom_Values```.  It seems like FrozenDictionary would immediately give us a speed boost simply by changing the dictionary type.

**Reply by LeaFrock** [reply]
You can change the type of `Custom_Values` to `JsonElement`，then convert it to frozen collection with [EnumerateArray()](https://learn.microsoft.com/zh-cn/dotnet/api/system.text.json.jsonelement.enumeratearray?view=net-7.0#system-text-json-jsonelement-enumeratearray) after deserialization.

**Reply by TonyValenti** [reply]
@LeaFrock If I understand you correctly, then I have to have two versions of ```PersonJson``` - one where ```Custom_Values``` is a ```JsonElement``` and another where it is a ```FrozenDictionary``` and then convert between the two.  Am I correct on that?

**Reply by LeaFrock** [reply]
You don't need to do so. One easy way I think of is something like the following,

```csharp
public record PersonJson : EntityJson { 
        public string? Name { get; init; }
        public JsonElement Custom_Values { get; init; } // <--- This is a high access property

        private FrozenDictionary<long, CustomValueJson> _dict;

        public FrozenDictionary<long, CustomValueJson> GetFrozenCustomValues()
        {
              if(_dict is null) // In cocurrent environment, use double-check lock or use Lazy<T>
              {
                    // initialize the frozen dict
              }
              return _dict;
        }
        
        // Or use an extra property with [JsonIgnore]
       [JsonIgnore]
       public FrozenDictionary<long, CustomValueJson> Custom_Values_Dict
       {
              get 
              {
                   // initialize and return the frozen dict
              }
       }
    }
```

It wouldn‘t require too much work for refactor.

**Reply by TonyValenti** [reply]
@LeaFrock - That would never pass code review.

I guess the question is whether we would write the deserializer ourselves or if it could be provided as part of STJ.  I hope that @stephentoub or another member will add it into STJ because (1) I imagine that caching deserialized Json is common and (2) I imagine the deserializer could share a lot of code with the ImmutableDictionary deserializer.
